src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -133,10 +138,15 @@ public TypeWithAnnotations GetInferredReturnType(ConversionsBase? conversions, N
                                        initialState: nullableState,
                                        returnTypes);
                 diagnostics.Free();
-                var inferredReturnType = InferReturnType(returnTypes, node: this, Binder, delegateType, Symbol.IsAsync, conversions);
+                inferredReturnType = InferReturnType(returnTypes, node: this, Binder, delegateType, Symbol.IsAsync, conversions);
                 returnTypes.Free();
-                return inferredReturnType.TypeWithAnnotations;
             }
+
+            if (!allowInferredFromFunctionType && inferredReturnType.InferredFromFunctionType)
+            {
+                return default;
```

Comment:
>return default; [](http://example.com/codeflow?start=16&length=15)

I would expect a much simpler change where a function type wouldn't be simply added to the set of candidates. That would revert behavior to C# 9 behavior, I think.